### PR TITLE
More scaladoc

### PR DESCRIPTION
**Pull request by djspiewak**
Still lots of stuff to do.

### CODE REVIEW

**Comment by vasilmkd** [comment]
If you are making more changes, tag me as a reviewer when you're done and I'll take a look. 😄

**Comment by vasilmkd** [comment]
I'll take a look later tonight.

File: core/shared/src/main/scala/cats/effect/IO.scala [hunk 1]
```diff
@@ -118,12 +120,23 @@ sealed abstract class IO[+A] private () extends IOPlatform[A] {
 
   /**
    * Runs the current IO, then runs the parameter, keeping its result.
-   * The result of the first action is ignored.
-   * If the source fails, the other action won't run.
+   * The result of the first action is ignored. If the source fails,
+   * the other action won't run. Not suitable for use when the parameter
+   * is a recursive reference to the current expression.
+   *
+   * @see [[>>]]
```
Comment (vasilmkd) [inline core/shared/src/main/scala/cats/effect/IO.scala]:
```suggestion
   * @see [[>>]] for the recursion-safe, lazily evaluated alternative
```

File: core/shared/src/main/scala/cats/effect/IO.scala [hunk 2]
```diff
@@ -103,6 +103,8 @@ import scala.util.{Failure, Success, Try}
  *       IO.pure(a)
  *   }
  * }}}
+ *
+ * @see [[IOApp]]
```
Comment (vasilmkd) [inline core/shared/src/main/scala/cats/effect/IO.scala]:
```suggestion
 * @see [[IOApp]] for the preferred way of executing whole programs wrapped in `IO`
```

File: core/shared/src/main/scala/cats/effect/IO.scala [hunk 3]
```diff
@@ -965,9 +1013,16 @@ object IO extends IOCompanionPlatform with IOLowPriorityImplicits {
...
    * An IO that contains an empty Option.
+   *
+   * @see [[some]]
    */
   def none[A]: IO[Option[A]] = pure(None)
 
+  /**
+   * An IO that contains some Option of the given value.
+   *
+   * @see [[none]]
```
Comment (vasilmkd) [inline core/shared/src/main/scala/cats/effect/IO.scala]:
```suggestion
   * @see [[none]] for the empty Option variant
```

File: core/shared/src/main/scala/cats/effect/IO.scala [hunk 4]
```diff
@@ -965,9 +1013,16 @@ object IO extends IOCompanionPlatform with IOLowPriorityImplicits {
 
   /**
    * An IO that contains an empty Option.
+   *
+   * @see [[some]]
```
Comment (vasilmkd) [inline core/shared/src/main/scala/cats/effect/IO.scala]:
```suggestion
   * @see [[some]] for the non-empty Option variant
```

File: core/shared/src/main/scala/cats/effect/IO.scala [hunk 5]
```diff
@@ -803,6 +843,14 @@ sealed abstract class IO[+A] private () extends IOPlatform[A] {
...
 
+  /**
+   * Evaluates the current `IO` in an infinite loop, terminating only on
+   * error or cancelation.
+   *
+   * {{{
+   *   IO.println("Hello, World!").foreverM    // continues printing forever
+   * }}}
+   */
   def foreverM: IO[Nothing] = Monad[IO].foreverM[A, Nothing](this)
```
Comment (vasilmkd) [inline core/shared/src/main/scala/cats/effect/IO.scala:860]: Does it matter if we change this to `IO.asyncForIO.foreverM[A, Nothing](this)`?

File: core/shared/src/main/scala/cats/effect/IO.scala [hunk 6]
```diff
@@ -603,7 +641,9 @@ sealed abstract class IO[+A] private () extends IOPlatform[A] {
 
   /**
    * Makes the source `IO` uninterruptible such that a [[cats.effect.kernel.Fiber#cancel]]
-   * signal has no effect.
+   * signal is ignored until completion.
+   *
+   * @see [[IO.uncancelable]]
```
Comment (vasilmkd) [inline core/shared/src/main/scala/cats/effect/IO.scala]:
```suggestion
   * @see [[IO.uncancelable]] for constructing uncancelable `IO` values with
   *                          user-configurable cancelable regions
```

File: core/shared/src/main/scala/cats/effect/IO.scala [hunk 7]
```diff
@@ -321,6 +350,15 @@ sealed abstract class IO[+A] private () extends IOPlatform[A] {
...
     IO.bracketFull(_ => this)(use)(release)
 
+  /**
+   * Shifts the execution of the current IO to the specified `ExecutionContext`.
+   * All stages of the execution will default to the pool in question, and any
+   * asynchronous callbacks will shift back to the pool upon completion. Any nested
+   * use of `evalOn` will override the specified pool. Once the execution fully
+   * completes, default control will be shifted back to the enclosing (inherited) pool.
+   *
+   * @see [[IO.executionContext]]
```
Comment (vasilmkd) [inline core/shared/src/main/scala/cats/effect/IO.scala]:
```suggestion
   * @see [[IO.executionContext]] for obtaining the `ExecutionContext` on which
   *                              the current `IO` is being executed
```

File: core/shared/src/main/scala/cats/effect/IO.scala [hunk 8]
```diff
@@ -118,12 +120,23 @@ sealed abstract class IO[+A] private () extends IOPlatform[A] {
...
   def *>[B](that: IO[B]): IO[B] =
     productR(that)
 
+  /**
+   * Runs the current IO, then runs the parameter, keeping its result.
+   * The result of the first action is ignored.
+   * If the source fails, the other action won't run. Evaluation of the
+   * parameter is done lazily, making this suitable for recursion.
+   *
+   * @see [*>]
```
Comment (vasilmkd) [inline core/shared/src/main/scala/cats/effect/IO.scala]:
```suggestion
   * @see [*>] for the strictly evaluated alternative
```

File: core/shared/src/main/scala/cats/effect/IO.scala [hunk 9]
```diff
@@ -170,6 +183,14 @@ sealed abstract class IO[+A] private () extends IOPlatform[A] {
   def option: IO[Option[A]] =
     redeem(_ => None, Some(_))
 
+  /**
+   * Runs the current and given IO in parallel, producing the pair of
+   * the outcomes. Both outcomes are produced, regardless of whether
+   * they complete successfully.
+   *
+   * @see [[both]]
```
Comment (vasilmkd) [inline core/shared/src/main/scala/cats/effect/IO.scala]:
```suggestion
   * @see [[both]] for the version which embeds the outcomes to produce a pair
   *               of the results
```

File: core/shared/src/main/scala/cats/effect/IO.scala [hunk 10]
```diff
@@ -170,6 +183,14 @@ sealed abstract class IO[+A] private () extends IOPlatform[A] {
...
+   * @see [[raceOutcome]]
```
Comment (vasilmkd) [inline core/shared/src/main/scala/cats/effect/IO.scala]:
```suggestion
   * @see [[raceOutcome]] for the version which produces the outcome of the
   *                      winner and cancels the loser of the race
```

File: core/shared/src/main/scala/cats/effect/IO.scala [hunk 11]
```diff
@@ -178,6 +199,14 @@ sealed abstract class IO[+A] private () extends IOPlatform[A] {
       }
     }
 
+  /**
+   * Runs the current and given IO in parallel, producing the pair of
+   * the results. If either fails with an error, the result of the whole
+   * will be that error and the other will be canceled.
+   *
+   * @see [[bothOutcome]]
```
Comment (vasilmkd) [inline core/shared/src/main/scala/cats/effect/IO.scala]:
```suggestion
   * @see [[bothOutcome]] for the version which produces the outcome of both
   *                      effects executed in parallel
```

File: core/shared/src/main/scala/cats/effect/IO.scala [hunk 12]
```diff
@@ -178,6 +199,14 @@ sealed abstract class IO[+A] private () extends IOPlatform[A] {
...
+   * @see [[race]]
```
Comment (vasilmkd) [inline core/shared/src/main/scala/cats/effect/IO.scala]:
```suggestion
   * @see [[race]] for the version which produces the result of the winner and
   *               cancels the loser of the race
```

File: build.sbt [hunk 13]
```diff
@@ -312,9 +312,7 @@ lazy val core = crossProject(JSPlatform, JVMPlatform)
       ProblemFilters.exclude[MissingClassProblem]("cats.effect.SyncIO$Delay"),
       ProblemFilters.exclude[DirectMissingMethodProblem]("cats.effect.IO#IOCont.apply"),
       ProblemFilters.exclude[DirectMissingMethodProblem]("cats.effect.IO#IOCont.copy"),
-      ProblemFilters.exclude[DirectMissingMethodProblem]("cats.effect.IO#IOCont.this")
-    )
-  )
+      ProblemFilters.exclude[DirectMissingMethodProblem]("cats.effect.IO#IOCont.this")))
```
Comment (vasilmkd) [inline build.sbt:315]: Do we want to run a proper `scalafmtSbt` for the whole project?

File: core/jvm/src/main/scala/cats/effect/IOApp.scala [hunk 14]
```diff
@@ -20,16 +20,193 @@ import scala.concurrent.{blocking, CancellationException}
...
+ *       for {
+ *         _ <- IO.print("Enter your name: ")
+ *         name <- IO.readln
+ *         _ <- IO.println("Hello, " + name)
+ *       } yield ExitCode.Success
+ *   }
+ * }}}
+ *
+ * In the above example, `MyApplication` will be a runnable class with
+ * a `main` method, visible to Sbt, IntelliJ, or plain-old `java`. When
```
Comment (vasilmkd) [inline core/jvm/src/main/scala/cats/effect/IOApp.scala:50]: Nitpicking, isn't `sbt` stylized in the lowercase form?

File: core/jvm/src/main/scala/cats/effect/IOApp.scala [hunk 15]
```diff
@@ -20,16 +20,193 @@ import scala.concurrent.{blocking, CancellationException}
 
 import java.util.concurrent.CountDownLatch
 
+/**
```
Comment (vasilmkd) [inline core/jvm/src/main/scala/cats/effect/IOApp.scala:23]: This is great.

**Review by vasilmkd — COMMENTED** [review]
Much appreciated scaladoc. I left a few suggestions, feel free to accept them or ignore them. I will officially Approve the PR before merging.

File: core/shared/src/main/scala/cats/effect/IO.scala (hunk 5 above)
Comment (djspiewak) [reply core/shared/src/main/scala/cats/effect/IO.scala:860]: It's probably slightly faster to do so in fact. Very slightly.

File: build.sbt (hunk 13 above)
Comment (djspiewak) [reply build.sbt:315]: Eh… We can. :-) Scalafmt doesn't handle the trailing parentheses case correctly though, so it wouldn't have found this one.

File: core/jvm/src/main/scala/cats/effect/IOApp.scala (hunk 14 above)
Comment (djspiewak) [reply core/jvm/src/main/scala/cats/effect/IOApp.scala:50]: Is it? I can't keep track tbh

**Review by vasilmkd — APPROVED** [review]

**Pull request merged**